src/storage/index/hash_index.cpp

```diff
@@ -363,6 +418,198 @@ void HashIndex<T>::copyEntryToSlot(slot_id_t slotId, const T& entry, uint8_t fin
+    // immediately to avoid increasing memory usage (caching one page of slots at a time since split
+    // slots usually get rehashed to a new page).
+    //
+    // On the other hand, two passes may not be significantly slower than one
+    // TODO: one pass would also reduce locking when frames are unpinned,
+    // which is useful if this can be paralellized
+    reserve(bulkInsertLocalStorage.size());
+    KU_ASSERT(
+        bulkInsertLocalStorage.pSlots->size() == pSlots->getNumElements(TransactionType::WRITE));
+    KU_ASSERT(this->indexHeaderForWriteTrx->currentLevel ==
+              bulkInsertLocalStorage.indexHeader.currentLevel);
+    KU_ASSERT(this->indexHeaderForWriteTrx->levelHashMask ==
+              bulkInsertLocalStorage.indexHeader.levelHashMask);
+    KU_ASSERT(this->indexHeaderForWriteTrx->higherLevelHashMask ==
+              bulkInsertLocalStorage.indexHeader.higherLevelHashMask);
+    KU_ASSERT(this->indexHeaderForWriteTrx->nextSplitSlotId ==
+              bulkInsertLocalStorage.indexHeader.nextSplitSlotId);
+
+    auto originalNumEntries = this->indexHeaderForWriteTrx->numEntries;
+
+    // Storing as many slots in-memory as on-disk shouldn't be necessary (for one it makes memory
+    // usage an issue as we may need significantly more memory to store the slots that we would
+    // otherwise) Instead, when merging here we can re-hash and split each in-memory slot (into
+    // temporary vector buffers instead of slots for improved performance) and then merge each of
+    // those one at a time into the disk slots. That will keep the low memory requirements and still
+    // let us update each on-disk slot one at a time.
+
+    auto diskSlotIterator = pSlots->iter();
+    // TODO: Use a separate random access iterator and one that's sequential for adding new overflow
+    // slots All new slots will be sequential and benefit from caching, but for existing randomly
+    // accessed slots we just benefit from the interface. However, the two iterators would not be
+    // able to pin the same page simultaneously
+    auto diskOverflowSlotIterator = oSlots->iter();
+    for (uint64_t slotId = 0; slotId < bulkInsertLocalStorage.pSlots->size(); slotId++) {
+        auto localSlot =
+            typename HashIndexBuilder<T>::SlotIterator(slotId, &bulkInsertLocalStorage);
+        // If mask is empty, skip this slot
+        if (!localSlot.slot->header.validityMask) {
+            // There should be no entries in the overflow slots, as we never leave gaps in the
+            // builder
+            continue;
+        }
+        diskSlotIterator.seek(slotId);
+        slot_id_t diskEntryPos = 0u;
+        auto* diskSlot = &*diskSlotIterator;
+        // Merge slot from local storage to existing slot
+        do {
+            for (auto entryPos = 0u; entryPos < getSlotCapacity<T>(); entryPos++) {
+                if (localSlot.slot->header.isEntryValid(entryPos)) {
```

Comment:
This piece of code presents an optimization opportunity that we should add a fast path for the case that all valid $k$ entries are stored as first $k$ ones in the slot. ( In many cases, there is no deletions to entries in local slots). so we can skip the redundant if check on validity bitmap, and also we don't need to loop until slot capacity, which may not make much difference.

Reply:
Yes, that's a good idea. I think when the builder/`InMemHashIndex` gets restructured the validity mask can be removed and we can just store the number of entries instead.